ksqldb-engine/src/main/java/io/confluent/ksql/util/BinPackedStreamsMetadata.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2021 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.util;
+
+import com.google.common.base.Throwables;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import io.confluent.ksql.name.SourceName;
+import io.confluent.ksql.query.KafkaStreamsBuilder;
+import io.confluent.ksql.query.QueryError;
+import io.confluent.ksql.query.QueryErrorClassifier;
+import io.confluent.ksql.query.QueryId;
+import io.confluent.ksql.rest.entity.StreamsTaskMetadata;
+import org.apache.kafka.streams.KafkaStreams;
+import org.apache.kafka.streams.LagInfo;
+import org.apache.kafka.streams.Topology;
+import org.apache.kafka.streams.errors.StreamsException;
+import org.apache.kafka.streams.errors.StreamsUncaughtExceptionHandler;
+import org.apache.kafka.streams.state.StreamsMetadata;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class BinPackedStreamsMetadata {
+
+  private static final Logger LOG = LoggerFactory.getLogger(BinPackedStreamsMetadata.class);
+
+  private final KafkaStreamsBuilder kafkaStreamsBuilder;
+  private QueryErrorClassifier errorClassifier ;
+  private KafkaStreams kafkaStreams;
+  private ImmutableMap<String, Object> streamsProperties;
+  private final QueryMetadataImpl.TimeBoundedQueue queryErrors;
+  private final Map<String, Topology> queryMap;
```

Comment:
Could we just maintain a single map with a consolidated class here? E.g. if we have a `BinPackedQuery` which contains `QueryId, Topology, BinPackedPersistentQueryMetadataImpl`. Then we can just maintain a single map here which is `<QueryId, BinPackedQuery>`. Then in the `BinPackedPersistentQueryMetadataImpl` above it only need to remember its `QueryId` and call `startQuery(queryId)` upon initialization.

Reply:
I  can probably make this simpler, and add some more getters to the `BinPackedPersistentQueryMetadataImpl` to fill in the gaps.